server/recurring/cluster.go

```diff
@@ -0,0 +1,271 @@
+}
+
+type PreProcessor struct {
+	documents []Document
+	// Word count
+	wc  *WordCount
+	idf map[string]float64
+}
+
+func (p *PreProcessor) AddTransaction(txn *Transaction) {
+	words := cleanStringRegex.FindAllString(txn.OriginalName, len(txn.OriginalName))
+	if txn.OriginalMerchantName != nil {
+		words = append(words, strings.Fields(*txn.OriginalMerchantName)...)
+	}
+
+	wordCounts := make(map[string]int, len(words))
+	// Get the term frequency from the transaction name
+	for _, word := range words {
+		word = strings.ToLower(word)
+		wordCounts[word]++
+		p.wc.Increment(word)
+	}
+
+	tf := make(map[string]float64, len(wordCounts))
+	for word, count := range wordCounts {
+		tf[word] = float64(count) / float64(len(words))
+	}
+
+	p.documents = append(p.documents, Document{
+		Transaction: txn,
+		TF:          tf,
+		TFIDF:       map[string]float64{},
+	})
+}
+
+// PostPrepareCalculations should be called after all transactions have been added to the dataset.
+func (p *PreProcessor) PostPrepareCalculations() {
+	docCount := float64(len(p.documents))
+	p.wc.Iterate(func(word string, _ int, count int) {
+		p.idf[word] = math.Log(docCount / (float64(count) + 1))
+	})
+	var min, max float64
+	for i := range p.documents {
+		document := p.documents[i]
+		for word, tfValue := range document.TF {
+			document.TFIDF[word] = tfValue * p.idf[word]
+		}
+		p.documents[i] = document
+		amount := float64(document.Transaction.Amount)
+		if amount > max || max == 0 {
+			max = amount
+		}
+		if amount < min || min == 0 {
+			min = amount
+		}
+	}
+
+	minified := p.wc.Minify()
+	for i := range p.documents {
+		document := p.documents[i]
+		amount := float64(document.Transaction.Amount)
+		adjusted := (amount - min) / (max - min)
+		document.AmountScaled = 2*adjusted - 1
+
+		// Calculate a vector based on all the words from all transactions.
+		// TODO Also need to look into Principal Component Analysis
+		vector := make([]float64, len(minified))
+		for word, tfidfValue := range document.TFIDF {
+			index, exists := minified[word]
+			if !exists {
+				continue
+			}
+			vector[index] = tfidfValue
+		}
+		var norm float64
+		for _, value := range vector {
+			norm += value * value
+		}
+		norm = math.Sqrt(norm)
+		for i, value := range vector {
+			vector[i] = value / norm
+		}
+		document.Vector = append([]float64{document.AmountScaled}, vector...)
+		p.documents[i] = document
+	}
+}
+
+func (p *PreProcessor) GetDatums() []Datum {
+	datums := make([]Datum, len(p.documents))
+	for i, document := range p.documents {
+		datums[i] = Datum{
+			ID:          document.Transaction.TransactionId,
+			Transaction: *document.Transaction,
+			Vector:      document.Vector,
+		}
+	}
+
+	return datums
+}
+
+type Datum struct {
+	ID          uint64
+	Transaction Transaction
+	Vector      []float64
+}
+
+func (a Datum) Distance(b Datum) float64 {
+	var distance float64
+	for i, value := range a.Vector {
+		distance += math.Pow(value-b.Vector[i], 2)
+	}
+	return distance
```

Comment:
Could also be done with SIMD

```asm
TEXT ·sumOfSquaredDifferences(SB), NOSPLIT, $0
    MOVQ a+0(FP), SI            // Load the address of a.Vector
    MOVQ b+24(FP), DX           // Load the address of b.Vector
    MOVQ a+8(FP), CX            // Load the length of the vectors
    VXORPD YMM0, YMM0, YMM0     // Clear the accumulator register YMM0

    // Loop over the vectors in batches of 4 (since we're using YMM registers)
    LOOP:
        VMOVUPD (SI), YMM1      // Load 4 float64 elements from a.Vector
        VMOVUPD (DX), YMM2      // Load 4 float64 elements from b.Vector
        VSUBPD YMM2, YMM1, YMM1 // Compute the difference (a-b)
        VMULPD YMM1, YMM1, YMM1 // Square the differences
        VADDPD YMM1, YMM0, YMM0 // Accumulate the squared differences

        ADDQ $32, SI            // Move to the next 4 elements in a.Vector
        ADDQ $32, DX            // Move to the next 4 elements in b.Vector
        SUBQ $4, CX             // Decrement the loop counter (4 elements processed)
        JNZ LOOP                // Continue loop if there are more elements

    // Finalize the sum of squared differences
    // Extract the high 128 bits and add them to the low 128 bits
    VEXTRACTF128 $0x1, YMM0, XMM1
    VADDSD XMM1, XMM0, XMM0
    VHADDPD XMM0, XMM0, XMM0

    // Store the result
    VMOVSD XMM0, ret+48(FP)

    RET
```

Reply:
A visualization of the registers:

```
 YMM0, YMM1, YMM2 register (256 bits total):
+------------------+------------------+------------------+------------------+
|  float64 value 0 |  float64 value 1 |  float64 value 2 |  float64 value 3 |
+------------------+------------------+------------------+------------------+
```

Reply:
instructions are backwards